Framework: Cleaning up layout styles

shaunandrews opened the pull request:
The Layout "module handles the base page layout for Calypso leveraged by all other
sub-views, including the WordPress.com masterbar." The CSS is a little messy and fragile.

This PR aims to make the CSS easier to understand and ensure smoother transitions between layout focus states. The only visual changes you should see are related to the transition of the site-selector and sidebar — the animations are the same, but they should be smoother.

This will affect every screen within Calypso and needs some thorough testing.

gwwar commented:
If we're okay with such a large refactor and will be doing full manual regression testing, would it be possible to also remove `backface-visibility: hidden` from [Masterbar](https://github.com/Automattic/wp-calypso/blob/a35865b125c7eac01725fa45da36dec9ab893d6f/client/layout/masterbar/style.scss#L32) and [Main](https://github.com/Automattic/wp-calypso/blob/a35865b125c7eac01725fa45da36dec9ab893d6f/client/components/main/style.scss#L5)? This CSS rule triggers a [compositing layer](https://href.li/?https://www.smashingmagazine.com/2016/12/gpu-animation-doing-it-right/) to be created in some browsers (hardware acceleration), this normally isn't an issue for small elements we animate, but these are pretty base nodes, meaning that any children or items imposed on top will also be hardware accelerated.

There are some tradeoffs in enabling HW acceleration, If there weren't, browsers would accelerate everything by default. 😄 

See p9oQ9f-23-p2 for more context.

gwwar commented on client/layout/style.scss:
```diff
@@ -72,135 +115,140 @@
+.layout.focus-content {
+	@include breakpoint( "<660px" ) {
+		.layout__secondary {
+			transform: translateX( -100% );
```
Is it possible to `display: none` instead? This would save the browser some work.

shaunandrews replied:
Yes, its possible — but then we lose the "sliding" transition.

gwwar commented on client/layout/style.scss:
```diff
@@ -72,135 +115,140 @@
-		@media print {
-			display: none;
+// Setup the site-selector element. Its default
+// position is off screen.
```
Why offscreen? Is there some JS that needs node dimensions? If not can we `display: none`?

shaunandrews replied:
If we `display: none;` then we can't do the transitions.

gwwar replied:
Feels like we should still be able to accomplish this, and not need to render misc elements. If it's too tricky with CSS, we can always do add a JS transition. CSS animation appears to set @nylen's comp on 🔥 anyway.

nylen replied:
For what it's worth, it's the continuous CSS animations that really slow things down for me.  Quick transitions are fine.  Though I agree we shouldn't render elements offscreen for transitions, instead we can use a bit of JS that renders the element when the transition should start.

gwwar replied:
I'll go ahead and defer on this for now. I suppose we might run into the trap of needing to add animation classes and then may get into buggy states if we fail to remove them.

gwwar commented on client/layout/style.scss:
```diff
@@ -1,10 +1,52 @@
+
+	// set a delay threshold for opacity changes
+	// prevents showing loader on fast connections
+	visibility: hidden;
```
Again is there a reason for not using `display: none`? Does the delay not work otherwise? I assume this is for the animation, but I'm thinking this might work if we add a` display:block` on the `is-active` modifier

shaunandrews replied:
This is the one part of the file I didn't touch! I'm not sure why its using `visibility` instead of `display`. I'll update it.

gwwar commented:
Noticed that the sidebar is triggering a loading element (that I usually never see) vs master.

gwwar commented on client/layout/style.scss:
```diff
@@ -72,135 +115,140 @@
+		left: 0;
+	pointer-events: none;
+	transform: translateX( -$sidebar-width-max );
+	pointer-events: none;
```
Duplicate `pointer-events` rule. (Note that i'm not a huge fan of modifying this value)

nylen replied:
+1 for not modifying `pointer-events`.

drw158 commented:
Looks good in Chrome on MacOS and Android

gwwar commented:
Verified that items in primary with a z-index don't leak through to sidebar #22478 due to the following rule adding a z-index stacking context to primary. Could we add an extra comment around this to avoid breakage if we refactor?
```
`transform: translateX(100%)`
```

![screen shot 2018-02-16 at 5 48 29 pm](https://user-images.githubusercontent.com/1270189/36336935-e8f48cfc-1341-11e8-9b31-083d0a909b5d.png)

![screen shot 2018-02-16 at 5 48 19 pm](https://user-images.githubusercontent.com/1270189/36336942-f5365ec8-1341-11e8-9f52-2a155c3143f2.png)

gwwar commented on client/layout/style.scss:
```diff
@@ -72,135 +115,140 @@
 		pointer-events: none;
+
+		@include breakpoint( "<660px" ) {
+			transform: translateX( 100% );
```
Sorry, should have made https://github.com/Automattic/wp-calypso/pull/22507#issuecomment-366406099 inline here :)

alisterscott commented:
Hi @gwwar 

I investigated why the full suite tests are failing and it looks like there's two issues on themes on this branch (which I can't reproduce on master)

The first issue is once you activate a theme and click "Learn More about this theme" a block where the sidebar was is displayed:

<img width="1446" alt="this pr" src="https://user-images.githubusercontent.com/128826/36358978-299a27ea-1562-11e8-9b05-10aa13070d66.png">

Steps to reproduce

1. Visit https://calypso.live/?branch=update/layout-styles (or http://calypso.localhost:3000 on this branch)
2. Choose your site
3. Select Themes
4. Choose a theme and click "Activate"
5. Click "Learn more about this theme"
(this also affects all sites - so select all sites, then a theme, then activate, then a site...)

![theme](https://user-images.githubusercontent.com/128826/36359008-546ba6ba-1562-11e8-8d09-521ef4e8ea1d.gif)

alisterscott commented:
The second theme issue is that when you're on the theme page and you switch from a single site to "all sites" (or vice-versa) the drop downs stop functioning.

Steps to reproduce:

1. Visit http://calypso.localhost:3000/themes/ and select a single site
2. Make sure elipisis buttons on themes work
3. Use "Switch Site" and select "All My Sites"
4. Try elipsis buttons - won't work
5. Refresh page - buttons now work
(When refreshing page on all sites on calypso.localhost it seems to make the screen poorly formatted - I believe this is calypso.localhost only and also this happens on `master`)

![themes switch](https://user-images.githubusercontent.com/128826/36359214-63f197c8-1564-11e8-9b51-90fd5ffebb9d.gif)

alisterscott commented:
> @alisterscott do you have confidence in our e2e coverage across different screen widths?

We run them at desktop and mobile widths to hit the breakpoints, and I'm confident we have functional coverage of the main flows - do buttons, links, fields etc. work and can customers complete their main flows. 

> I'm wondering if we need to step back and improve e2e tests first before trying to land something like this.

The e2e tests aren't really for visual regression testing, we have visual regression tests, I'm not sure how we can utilise these for branches though. @Automattic/flowpatrol ?

nylen commented on client/components/site-selector/style.scss:
```diff
@@ -5,15 +5,13 @@
 		display: flex;
 		position: relative;
 		opacity: 1;
-		pointer-events: initial;
```
Because of this removal, I'm unable to click on the site search bar after clicking Switch Site.  Elsewhere in this file there is a `pointer-events: none` rule for this element, and it is no longer being overridden.  https://github.com/Automattic/wp-calypso/pull/22507/files#diff-82484b636bfcea047c5bc590322bc7c4L104

shaunandrews replied:
Ah, nice catch. I've no clue why we're toying with pointer-events here — I've removed it and everything seem to work as expected.

nylen commented:
In this branch I am sometimes unable to click the white overlay to dismiss the site selector.  I'm not sure exactly what is causing this but it happened a few times when I resized across breakpoints with the site selector open.

hoverduck commented:
> The e2e tests aren't really for visual regression testing, we have visual regression tests, I'm not sure how we can utilise these for branches though. @Automattic/flowpatrol ?

We can discuss this on Slack/P2 to try and figure out what it would take and what we'd get out of it, but offhand I don't think it would be worthwhile...at least not any time soon.  We can add it to the "hopefully someday" list :smile:

shaunandrews commented:
@alisterscott The /theme related things should be resolved now. Mind taking another look?

alisterscott commented:
@shaunandrews thanks for the fixes - the issues I mentioned previously have been resolved.

I did find some new issues with the updates though.

1) When using site switcher the "All sites" element is only partly selectable - it only highlights when chosen from the right hand side - this GIF shows this:

![site switcher](https://user-images.githubusercontent.com/128826/36469617-89232e28-1733-11e8-9759-0e131381028e.gif)

2) When using the site selector on the theme preview dialog when on /themes whilst selecting all sites  - I had trouble switching between sites - this may be related to 1 above?

![theme all sites](https://user-images.githubusercontent.com/128826/36469647-ada93f94-1733-11e8-9aa9-5187104e88f7.gif)

Thanks

alisterscott commented:
The e2e full suite tests highlighed the second issue

shaunandrews commented:
> When using site switcher the "All sites" element is only partly selectable - it only highlights when chosen from the right hand side - this GIF shows this:

I pushed a fix for this. The search input was the culprit — its always there, just doesn't display until you start typing. My fix applies `pointer-events: none;` to the input unless its parent element has the `.is-large` class.

> When using the site selector on the theme preview dialog when on /themes whilst selecting all sites - I had trouble switching between sites - this may be related to 1 above?

This is a weird one. I'm able to reproduce it on `master` in Chrome, so it doesn't appear to be a regression in this branch. Can you double check and confirm, @alisterscott ?

> I think it's just because this PR hasn't been rebased

Correct. I'm terrified to rebase — I've never done it without breaking everything. I'll ask for help :)

> The first time I opened an individual theme page it was missing the masterbar styles:

This seems to be an issue with some server-side rendering magic we do on `/themes` and only appears to affect `calypso.live` and `calypso.localhost`.

lsinger commented:
Rebased. 🤞

gwwar commented on client/components/main/style.scss:
```diff
@@ -28,7 +27,6 @@
 
 	@include breakpoint( "<660px" ) {
-		backface-visibility: hidden;
 		perspective: 1000;
```
Let's also try removing perspective 👍 This rule doesn't make sense outside of 3D elements.

gwwar commented:
1. Verified that Site Selection "All sites" looks good.
2. Appears to be a design flaw? A SitesDropdown component from inside of a modal doesn't have enough room to display results. I see the same behavior in prod, so I wouldn't block on this one.

alisterscott commented:
All tests passed ✅ 

Both regressions I mentioned fixed - looks good to me 

I agree with @gwwar that we should try to 🚢 this

shaunandrews commented:
Nothing like a few last minute tweaks to make sure everything goes smooth on merge. 😜 

I dropped the `perspective` attribute @gwwar mentioned, as it doesn't actually seem to be doing anything.

Also, I fixed a small overflow bug on the Add button in the Media sidebar menu item. It had `overflow: visible;` which caused a horizontal scroll on mobile. I changed it to `overflow: none;` and no more horizontal scrolling.

alisterscott commented:
All e2e functional tests still passing 👍

lancewillett commented:
Nice work @shaunandrews — thanks for sticking with this one.
